File: DiscriminatedUnions/Examples.ts
// Discriminated union in TypeScript.
// Це сукупність типів, де кожне значення має тег, який визначає різновид.

type Circle = { kind: "circle"; radius: number };
type Rectangle = { kind: "rectangle"; width: number; height: number };
type Triangle = { kind: "triangle"; baseLength: number; height: number };

type Shape = Circle | Rectangle | Triangle;

function area(shape: Shape): number {
  switch (shape.kind) {
    case "circle":
      return Math.PI * shape.radius * shape.radius;
    case "rectangle":
      return shape.width * shape.height;
    case "triangle":
      return (shape.baseLength * shape.height) / 2;
    default:
      const _exhaustiveCheck: never = shape;
      return _exhaustiveCheck;
  }
}

function describe(shape: Shape): string {
  switch (shape.kind) {
    case "circle":
      return `Circle with radius ${shape.radius}`;
    case "rectangle":
      return `Rectangle ${shape.width}x${shape.height}`;
    case "triangle":
      return `Triangle base ${shape.baseLength} height ${shape.height}`;
    default:
      const _exhaustiveCheck: never = shape;
      return _exhaustiveCheck;
  }
}

const shapes: Shape[] = [
  { kind: "circle", radius: 3 },
  { kind: "rectangle", width: 4, height: 5 },
  { kind: "triangle", baseLength: 6, height: 4 },
];

for (const shape of shapes) {
  console.log(`${describe(shape)} -> area = ${area(shape).toFixed(2)}`);
}

File: DiscriminatedUnions/Examples.fs
// Discriminated union in F#
// Це тип, який може бути одним із кількох варіантів, кожен з яких може мати свої дані.

type Shape =
    | Circle of radius: float
    | Rectangle of width: float * height: float
    | Triangle of baseLength: float * height: float

let area shape =
    match shape with
    | Circle radius -> System.Math.PI * radius * radius
    | Rectangle (width, height) -> width * height
    | Triangle (baseLength, height) -> baseLength * height / 2.0

let describe shape =
    match shape with
    | Circle radius -> sprintf "Circle with radius %.2f" radius
    | Rectangle (width, height) -> sprintf "Rectangle %.2fx%.2f" width height
    | Triangle (baseLength, height) -> sprintf "Triangle base %.2f height %.2f" baseLength height

[<EntryPoint>]
let main _ =
    let shapes = [
        Circle 3.0
        Rectangle (4.0, 5.0)
        Triangle (6.0, 4.0)
    ]

    shapes
    |> List.iter (fun shape ->
        printfn "%s -> area = %.2f" (describe shape) (area shape))

    0
